<!DOCTYPE html>
<html>
  <head>
    <meta http-eqiv="content-type" content="text/html;charset=utf-8">
    <title>Go con Ejemplos: Panic</title>
    <link rel=stylesheet href="site.css">
  </head>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-51269806-1', 'goconejemplos.com');
    ga('send', 'pageview');

  </script>
  <body>
    <div class="example" id="panic">
      <h2><a href="./">Go con Ejemplos</a>: Panic</h2>
      
      <table>
        
        <tr>
          <td class="docs">
            <p>Una llamada a <code>panic</code> típicamente significa que sucedió
un error inesperado. Lo usamos principalmente para terminar
la ejecución en caso de errores que no debieran aparecer
durante la operación normal o que no estamos listos
para manejar adecuadamente.</p>

          </td>
          <td class="code empty leading">
          
            
          </td>
        </tr>
        
        <tr>
          <td class="docs">
            
          </td>
          <td class="code leading">
          <a href="http://play.golang.org/p/RL5PNBD-KU"><img title="Run code" src="play.png" class="run" /></a>
            <div class="highlight"><pre><span class="kn">package</span> <span class="nx">main</span>
</pre></div>

          </td>
        </tr>
        
        <tr>
          <td class="docs">
            
          </td>
          <td class="code leading">
          
            <div class="highlight"><pre><span class="kn">import</span> <span class="s">&quot;os&quot;</span>
</pre></div>

          </td>
        </tr>
        
        <tr>
          <td class="docs">
            
          </td>
          <td class="code leading">
          
            <div class="highlight"><pre><span class="kd">func</span> <span class="nx">main</span><span class="p">()</span> <span class="p">{</span>
</pre></div>

          </td>
        </tr>
        
        <tr>
          <td class="docs">
            <p>Usaremos <code>panic</code> a lo largo de este sitio para comprobar
errores inesperados. Este el el único programa en el
sitio diseñado para llamar a la función <code>panic</code></p>

          </td>
          <td class="code leading">
          
            <div class="highlight"><pre>    <span class="nb">panic</span><span class="p">(</span><span class="s">&quot;un problema&quot;</span><span class="p">)</span>
</pre></div>

          </td>
        </tr>
        
        <tr>
          <td class="docs">
            <p>Un uso común de <code>panic</code> es abortar la ejecución
si una función devuelve un valor de error que no
sabemos (o queremos) manejar. Este es un ejemplo
de <code>panic</code> si encontramos un error inesperado al
momento de crear un archivo nuevo</p>

          </td>
          <td class="code">
          
            <div class="highlight"><pre>    <span class="nx">_</span><span class="p">,</span> <span class="nx">err</span> <span class="o">:=</span> <span class="nx">os</span><span class="p">.</span><span class="nx">Create</span><span class="p">(</span><span class="s">&quot;/tmp/file&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nx">err</span> <span class="o">!=</span> <span class="kc">nil</span> <span class="p">{</span>
        <span class="nb">panic</span><span class="p">(</span><span class="nx">err</span><span class="p">)</span>
    <span class="p">}</span>
<span class="p">}</span>
</pre></div>

          </td>
        </tr>
        
      </table>
      
      <table>
        
        <tr>
          <td class="docs">
            <p>Ejecutar este programa provocará la llamada a panic,
imprimirá un mensaje de error, el registro de ejecución
de la goroutine y terminará con un estatus diferente a
cero.</p>

          </td>
          <td class="code leading">
          
            <div class="highlight"><pre><span class="gp">$</span> go run panic.go
<span class="go">panic: un problema</span>
</pre></div>

          </td>
        </tr>
        
        <tr>
          <td class="docs">
            
          </td>
          <td class="code leading">
          
            <div class="highlight"><pre><span class="go">goroutine 1 [running]:</span>
<span class="go">main.main()</span>
<span class="go">	/.../panic.go:12 +0x47</span>
<span class="go">...</span>
<span class="go">exit status 2</span>
</pre></div>

          </td>
        </tr>
        
        <tr>
          <td class="docs">
            <p>Observe que a diferencia de algunos lenguajes que
utilizan excepciones para el manejo de errores, en Go
es propio del lenguaje devolver valores de error cada
vez que sea posible</p>

          </td>
          <td class="code empty">
          
            
          </td>
        </tr>
        
      </table>
      
      
      <p class="next">
        Siguiente ejemplo: <a href="defer">Defer</a>.
      </p>
      
      <p class="footer">
        by <a href="https://twitter.com/mmcgrana">@mmcgrana</a> | <a href="mailto:mmcgrana@gmail.com">feedback</a> | <a href="https://github.com/mmcgrana/gobyexample/blob/master/examples/panic">source</a> | <a href="https://github.com/mmcgrana/gobyexample#license">license</a>
      </p>
    </div>
  </body>
</html>
